File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        ans = 0;
        findHeight(root);
        return ans;
    }

    public int findHeight(TreeNode root) {
        if(root==null){
            return 0;
        }

        int leftTreeHeight = findHeight(root.left);
        int rightTreeHeight = findHeight(root.right);

        int diameter = leftTreeHeight + rightTreeHeight;

        ans = Math.max(ans, diameter);

        return Math.max(leftTreeHeight,rightTreeHeight) + 1;        
    }
}

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.findHeight(root)

        return self.ans
    
    def findHeight(self, root):
        # base
        if not root:
            return 0
        
        # logic
        # count left and right at each level
        # sum of them is the answer
        # but in the reverse approach
        leftH = self.findHeight(root.left)
        
        rightH = self.findHeight(root.right)
        
        self.ans = max(self.ans, leftH+rightH)
        
        # while returning from the bottom level to top level
        # return the max of either one of the tree since we need to use deepest branch
        # addition 1 for that level
        return max(leftH, rightH) + 1
        